Firestore/src/Query.php

```diff
@@ -517,93 +529,238 @@ public function endBefore(array $fieldValues)
                         $this->valueMapper,
                         $collection
                     );
 
+                    $name = $this->childPath($collection, $value);
                     $value = new DocumentReference(
                         $this->connection,
                         $this->valueMapper,
-                        $c,
-                        $this->childPath($collection, $value)
+                        $parent,
+                        $name
                     );
-                } elseif ($value instanceof DocumentReference) {
-                    $name = $value->name();
+                } else {
+                    if ($value instanceof DocumentReference) {
+                        $name = $value->name();
+                        $parent = $value->parent()->name();
+                    } else {
+                        throw new \InvalidArgumentException(
+                            'The corresponding value for DOCUMENT_ID must be a ' .
+                            'string or a DocumentReference.'
+                        );
+                    }
+
                     if (!$this->isPrefixOf($collection, $name)) {
-                        throw new \BadMethodCallException(sprintf(
+                        throw new \InvalidArgumentException(sprintf(
                             '%s is not a part of the query result set and ' .
                             'cannot be used as a query boundary',
                             $name
                         ));
                     }
-                } else {
-                    throw new \BadMethodCallException(
-                        'The corresponding value for DOCUMENT_ID must be a ' .
-                        'string or a DocumentReference.'
-                    );
+
+                    if ($parent !== $collection) {
+                        throw new \InvalidArgumentException(
+                            'Only direct children may be used as query boundaries.'
+                        );
+                    }
                 }
+            }
+        }
 
-                if ($value->parent()->name() !== $collection) {
-                    throw new \BadMethodCallException(
-                        'Only direct children may be used as query boundaries.'
-                    );
+        return $this->newQuery([
+            $key => [
+                'before' => $before,
+                'values' => $this->valueMapper->encodeValues($fieldValues)
+            ],
+            'orderBy' => $orderBy
+        ], true);
+    }
+
+    /**
+     * Build cursors for document snapshots.
+     *
+     * @param DocumentSnapshot $snapshot The document snapshot
+     * @return array[] A list, where position 0 is fieldValues and position 1 is orderBy.
+     */
+    private function snapshotPosition(DocumentSnapshot $snapshot, array $orderBy)
+    {
+        $appendName = true;
+        $direction = self::DIR_ASCENDING;
+        $fieldValues = [];
```

Comment:
Similarly to `appendName` I don't think we need to default this, it looks like we can use the assignment below